src/ovirtapi/transform.js

```diff
@@ -661,10 +655,79 @@ const StorageDomain = {
   toInternal ({ storageDomain }: { storageDomain: ApiStorageDomainType }): StorageDomainType {
     const permissions = Permissions.toInternal({ permissions: storageDomain?.permissions?.permission })
 
+    // frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/dataprovider/AsyncDataProvider.java#getDiskVolumeFormat
+    const diskAttributesForDiskType: {
+      ['thin' | 'pre']: { sparse: boolean, format?: 'raw' | 'cow' }
+    } = {
+      thin: { sparse: true },
+      pre: { sparse: false },
+    }
+
+    // frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/dataprovider/AsyncDataProvider.java#getVolumeType
+    const defaultDiskFormatToSparse: {
+      ['raw' | 'cow']: boolean
+    } = {
+      cow: true,
+      raw: true,
+    }
+
+    const storageType = storageDomain?.storage?.type ?? 'unknown'
+    let storageSubType
+    switch (storageType) {
+      case 'nfs':
+      case 'localfs':
+      case 'posixfs':
+      case 'glusterfs':
+      case 'glance':
+        storageSubType = 'file'
+        diskAttributesForDiskType.thin.format = 'raw'
+        diskAttributesForDiskType.pre.format = 'raw'
+        defaultDiskFormatToSparse.cow = true
+        defaultDiskFormatToSparse.raw = true
+        break
+
+      case 'fcp':
+      case 'iscsi':
+        storageSubType = 'block'
+        diskAttributesForDiskType.thin.format = 'cow'
+        diskAttributesForDiskType.pre.format = 'raw'
+        defaultDiskFormatToSparse.cow = true
+        defaultDiskFormatToSparse.raw = false
+        break
+
+      case 'cinder':
+      case 'managed_block_storage':
+        storageSubType = 'openstack'
+        diskAttributesForDiskType.thin.format = undefined
+        diskAttributesForDiskType.pre.format = 'raw'
+        defaultDiskFormatToSparse.cow = true
```

Comment:
1. should this  mapping be symmetric to `diskAttributesForDiskType`?  i.e. if `thin` is not supported (`thin.format = undefined` then `...ToSparse == false`for both `cow/raw`
2. perhaps we could get rid of `defaultDiskFormatToSparse` and rely only on supported formats defined via `diskAttributesForDiskType` ?

Reply:
It is MUCH easier to have both already calculated and use them as needed.

A new disk is driven by disk type == thin | pre.

A template disk is driven by storage allocation and disk format.  The whole concept of disk type as thin or pre is not present at all on the Resource Allocation tab in webadmin:

Clone:
![Screenshot 2021-09-02 at 12-31-53 oVirt Open Virtualization Manager](https://user-images.githubusercontent.com/3985964/131882379-537bcf68-ba96-4fa5-8063-b817d8eb6e3c.png)

Thin:
![Screenshot 2021-09-02 at 12-33-05 oVirt Open Virtualization Manager](https://user-images.githubusercontent.com/3985964/131882525-fb61e17c-b2fb-4898-a692-022f7a7335c9.png)

Keeping those logic paths separate was the only way to these things "simple".

Reply:
>  The whole concept of disk type as thin or pre is not present at all on the Resource Allocation tab in webadmin
>

Webadmin is inconsistent here. "Create a disk from template for new VM" scenario has different design then "Add/Edit disk to existing VM" . 
First, I think that Resource Allocation tab **should** display the state of the `sparse` flag to the user. We seem to have it available and it is presented later on in Edit Disk dialog(screenshot below).
Second, `thin|clone` on this screens seem to map to `clone` flag which is independent from `sparse` flag (although sounds similar).


![image](https://user-images.githubusercontent.com/64194103/132330855-6473d396-800b-4775-b91e-c33a0fcefd5f.png)

Reply:
> > > 1. should this mapping be symmetric to diskAttributesForDiskType? i.e. if thin is not supported (thin.format = undefined then ...ToSparse == falsefor both cow/raw
> > 
> > 
> > undefined maps to Unassigned in engine
> 
>   1. So `thin.format == undefined == unassigned`. What does it mean from end user perspective? Is `thin` then a legal value to chose ?

I assume thin remains a legal value to choose, but I don't know for sure.  Worst case from web-ui is that an admin setups up a template with disks on an exotic storage domain, a user attempts to create a VM from the template, and the REST action fails.  

> 
>   2. we use the reverse mapping  only in the template scenario and always with `cow` format. So effectively `thin.format == cow` for this case.  How about making that difference explicit and providing different defaults for template vs new disk? i.e.  `newDisk.thin.format == undefined` vs `templateDisk.thin.format == cow`/
> 

In the template scenario, the disks do not always have `cow` format.  The format varies.  I tried to capture the permutations in the doc block for `determineTemplateDiskFormatAndSparse()`.

As mentioned and discusses in another comment thread, the disk type (thin | pre) doesn't matter in the template scenario.  Only the format & sparse matter.  There is no symmetric translation.

>   3. It seems we need some king of introduction/preamble to this code that would explain the convention. It also should get reviewed by the Storage team.
> 
> 
> > It is MUCH easier to have both already calculated and use them as needed.
> 
> Sure. However this obfuscates the mapping. IMHO `templateDisk.thin.format == cow` is more readable then `defaultDiskFormatToSparse.cow = true`
> 
> > The whole concept of disk type as thin or pre is not present at all on the Resource Allocation tab in webadmin
> 
> Webadmin is inconsistent here. "Create a disk from template for new VM" scenario has different design then "Add/Edit disk to existing VM" .

Yes it is.  The whole idea is to copy how webadmin does things "by default" for template disks.

> First, I think that Resource Allocation tab **should** display the state of the `sparse` flag to the user. We seem to have it available and it is presented later on in Edit Disk dialog(screenshot below).

Seems reasonable as it is very confusing that the only place disk format is visible is there.  But I get the feeling that it needs to be there for reasons that no one felt compelled to document in code.

> Second, `thin|clone` on this screens seem to map to `clone` flag which is independent from `sparse` flag (although sounds similar).
> 

A storage allocation of "Clone" will impact the disk format sometimes and may do different things to the disks that may not be obvious (or even visible) in the Disk dialog.  From my testing and reading of webadmin code, the important thing is to get the disk format correct to enable the underlying disk and disk attachment work to be done correctly.

Reply:
I added a whole bunch of comments to `vmChanges.js` to try to better explain what is going on with template disks.

Also keep in mind, the storage allocation of thin vs clone is really "sparse qcow backed by template disk" vs "fully independent clone of the template disk".  New disks will never be backed by another disk.